### PR TITLE
Agn 39 - Order management

### PR DESCRIPTION
**Pull request by CD10h**

### CODE REVIEW

File: geles-frontend/package-lock.json [hunk 1]
```diff
@@ -1,6 +1,4014 @@
 {
+  "name": "geles-frontend",
```
Comment (CD10h) [inline geles-frontend/package-lock.json:2]: accidental commit

**Review by titas251 — APPROVED** [review]
Good

File: geles-frontend/src/App.svelte [hunk 2]
```diff
@@ -78,17 +77,23 @@
...
+            <Route path="/update/:id" component={UpdateFlower} />
+            <Route path="/favorite/stats" component={FavoriteFlowersStats} />
+          {:else}
+            <Route path="/cart" component={Cart} />
+            <Route path="/flowers/favorite" component={FavoriteFlowers} />
+            <Route path="/order/:cartId" component={Order} />
+          {/if}
+          
+          <Route path="/orders" component={ManageOrders} />
+          <Route path="/editOrder/:id" component={EditOrder}/>
```
Comment (emiliskiskis) [inline geles-frontend/src/App.svelte:92]: Could be /order/edit/:id

File: geles-frontend/src/App.svelte [hunk 3]
```diff
@@ -78,17 +77,23 @@
...
         {/if}
         <Route component={NotFound} />
       </div>
     </Router>
   {/if}
+
```
Comment (emiliskiskis) [inline geles-frontend/src/App.svelte:98]: Unnecessary whitespace

File: geles-frontend/src/EditOrder.svelte [hunk 4]
```diff
@@ -0,0 +1,517 @@
...
+    version: 0
+  };
+
+  let editedUser: User = {
+    id: 0,
+    username: "",
+    photo: null,
+    admin: false,
+    cartId: 0
+  };
```
Comment (emiliskiskis) [inline geles-frontend/src/EditOrder.svelte:58]: Keep `let editedUser: User;` or `let editedUserName: string;`, since it's the only piece of data of user info used, it's safe to discard the rest

File: geles-frontend/src/EditOrder.svelte [hunk 5]
```diff
@@ -0,0 +1,517 @@
...
+  let flowers: Flower[];
+  let orderTotal: number = 0;
+
+  function flowerFromId(id: number): Flower {
+    return flowers.find(f => f.id == id) as Flower;
+  }
```
Comment (emiliskiskis) [inline geles-frontend/src/EditOrder.svelte:64]: Must check if undefined, avoid casts at all costs (rhyme intended)

File: geles-frontend/src/EditOrder.svelte [hunk 6]
```diff
@@ -0,0 +1,517 @@
...
+    try {
+      const response = await axios.get<User>("/users/", {
+        withCredentials: true
+      });
+      oldUser = response.data;
+    } catch (e) {
+      oldUser = null;
+    }
+    user.set(oldUser);
+  }
```
Comment (emiliskiskis) [inline geles-frontend/src/EditOrder.svelte:77]: Misplaced user store setter, refresh component based on App.svelte user store update

File: geles-frontend/src/EditOrder.svelte [hunk 7]
```diff
@@ -0,0 +1,517 @@
...
+
+  function recalcSum() {
+    orderTotal = order.orderFlowers
+      .map(
+        flowerInOrder =>
+          flowerFromId(flowerInOrder.flowerId).price * flowerInOrder.quantity
```
Comment (emiliskiskis) [inline geles-frontend/src/EditOrder.svelte:83]: Why not use `rowSum` function here?

File: geles-frontend/src/EditOrder.svelte (hunk 5 above)
Comment (emiliskiskis) [reply geles-frontend/src/EditOrder.svelte:64]: Also unnecessary function return type

File: geles-frontend/src/EditOrder.svelte [hunk 8]
```diff
@@ -0,0 +1,517 @@
...
+  }
+
+  function rowSum(flower: OrderFlower) {
+    return flower.quantity * flowerFromId(flower.flowerId).price;
+  }
+
+  function disableButton(
+    status: OrderStatus,
+    button: OrderEditButton
+  ): boolean {
```
Comment (emiliskiskis) [inline geles-frontend/src/EditOrder.svelte:102]: Unnecessary function return type

File: geles-frontend/src/EditOrder.svelte [hunk 9]
```diff
@@ -0,0 +1,517 @@
...
+        " " +
+        reason.response.headers;
+    } else if (reason.request) {
+      let errString = reason.request;
+    } else {
+      // Something happened in setting up the request that triggered an Error
+      let errString = reason;
+    }
+    console.log("Error", reason);
+    addNotification(errString, AppNotificationType.DANGER);
```
Comment (emiliskiskis) [inline geles-frontend/src/EditOrder.svelte:169]: Error notification shows ALL RESPONSE HEADERS???

File: geles-frontend/src/EditOrder.svelte [hunk 10]
```diff
@@ -0,0 +1,517 @@
...
+      .then(() => addNotification("Užsakymas patvirtintas",AppNotificationType.SUCCESS))
+      .catch(handleError);
+  }
+  function handleConfirmDelivered() {
+    axios
+      .post(`/orders/${id}/confirmDelivery`, { version: order.version })
+      .then(getOrderData)
+      .then(() => addNotification("Užsakymo pristatymas patvirtintas",AppNotificationType.SUCCESS))
+      .catch(handleError);
+  }
```
Comment (emiliskiskis) [inline geles-frontend/src/EditOrder.svelte:217]: Make all functions async

File: geles-frontend/src/EditOrder.svelte [hunk 11]
```diff
@@ -0,0 +1,517 @@
...
+      month = "" + (d.getMonth() + 1),
+      day = "" + d.getDate(),
+      year = d.getFullYear();
+
+    if (month.length < 2) month = "0" + month;
+    if (day.length < 2) day = "0" + day;
+
+    return [year, month, day].join("-");
+  }
+  function orderStatusString(status: any): string {
```
Comment (emiliskiskis) [inline geles-frontend/src/EditOrder.svelte:230]: Avoid `any` in this case, also unnecessary function return type

File: geles-frontend/src/EditOrder.svelte [hunk 12]
```diff
@@ -0,0 +1,517 @@
...
+  }
+  function handleConfirmDelivered() {
+    axios
+      .post(`/orders/${id}/confirmDelivery`, { version: order.version })
+      .then(getOrderData)
+      .then(() => addNotification("Užsakymo pristatymas patvirtintas",AppNotificationType.SUCCESS))
+      .catch(handleError);
+  }
+
+  function formatDate(date: any) {
```
Comment (emiliskiskis) [inline geles-frontend/src/EditOrder.svelte:219]: Avoid `any` in this case

File: geles-frontend/src/EditOrder.svelte [hunk 13]
```diff
@@ -0,0 +1,517 @@
...
+      default:
+        return "???????";
+    }
+  }
+
+  // Run code on component mount (once)
+  onMount(async () => {
+    await axios.get(`/flowers/`).then(resp => {
+      flowers = resp.data;
+    });
```
Comment (emiliskiskis) [inline geles-frontend/src/EditOrder.svelte:251]: This `await` is unnecessary, remove `await` and use assignment operator and two statements.

File: geles-frontend/src/EditOrder.svelte [hunk 14]
```diff
@@ -0,0 +1,517 @@
...
+              {#if order.orderStatus == "UNPAID"}
+                <input
+                  type="number"
+                  bind:value={flower.quantity}
+                  min="1"
+                  max="100"
+                  size="5"
+                  on:input={e => {
+                    recalcSum();
+                  }}
```
Comment (emiliskiskis) [inline geles-frontend/src/EditOrder.svelte:298]: `e` is not used, rewrite as `on:input={recalcSum}`

File: geles-frontend/src/EditOrder.svelte [hunk 15]
```diff
@@ -0,0 +1,517 @@
...
+          disabled={disableButton(order.orderStatus, OrderEditButton.PAY)}
+          on:click={() => handlePay()}
+        >
+          Apmokėti
+        </button>
+
+        <button
+          class="button cancelbutton"
+          disabled={disableButton(order.orderStatus, OrderEditButton.CANCEL)}
+          on:click={() => handleCancel()}
```
Comment (emiliskiskis) [inline geles-frontend/src/EditOrder.svelte:333]: `() => fn()` is unnecessary, use `fn`, e.g. `on:click={handleUpdate}`

File: geles-frontend/src/EditOrder.svelte [hunk 16]
```diff
@@ -0,0 +1,517 @@
...
+        <button
+          class="button paybutton"
+          disabled={disableButton(order.orderStatus, OrderEditButton.PAY)}
+          on:click={() => handlePay()}
+        >
+          Apmokėti
+        </button>
+
+        <button
+          class="button cancelbutton"
```
Comment (emiliskiskis) [inline geles-frontend/src/EditOrder.svelte:331]: If buttons already have `button` class, seems unnecessary to write `button` in front of modifier classes

File: geles-frontend/src/EditOrder.svelte [hunk 17]
```diff
@@ -0,0 +1,517 @@
...
+          </button>
+        </div>
+        <div class="editorder-inputrow">
+          <button
+            class="button"
+            disabled={disableButton(
+              order.orderStatus,
+              OrderEditButton.ADMIN_CONFIRMDELIVERED
+            )}
+            on:click={() => handleConfirmDelivered()}
```
Comment (emiliskiskis) [inline geles-frontend/src/EditOrder.svelte:430]: `() => fn()` is unnecessary, use `fn`, e.g. `on:click={handleUpdate}`

File: geles-frontend/src/EditOrder.svelte [hunk 18]
```diff
@@ -0,0 +1,517 @@
...
+  }
+
+  .editorder-textoutput {
+    display: table-cell;
+    padding-left: 30px;
+  }
+
+  label {
+    display: table-cell;
+  }
```
Comment (emiliskiskis) [inline geles-frontend/src/EditOrder.svelte:516]: Frowned upon table layout disguised as div layout?

File: geles-frontend/types/flower.d.ts [hunk 19]
```diff
@@ -26,9 +26,11 @@ interface FlowerInCart {
...
 enum OrderStatus {
-  UNPAID,
-  PAID,
-  DELIVERED
+  UNPAID = "UNPAID",
+  PAID = "PAID",
+  CONFIRMED = "CONFIRMED",
+  DELIVERED = "DELIVERED",
+  CANCELED = "CANCELED"
 }
```
Comment (emiliskiskis) [inline geles-frontend/types/flower.d.ts:34]: Remove enum from here

File: geles_backend/src/main/java/lt/aigen/geles/controller/OrdersController.java [hunk 20]
```diff
@@ -82,29 +82,180 @@ public OrdersController(CartRepository cartRepository, OrderRepository orderRepo
...
+                order.getOrderStatus().equals(Order.OrderStatus.CANCELED) )
+            return new ResponseEntity<>(HttpStatus.FORBIDDEN);
+        order.setOrderStatus(Order.OrderStatus.CANCELED);
+
+        //check manually because DTO does not map to order
+        if (!orderVersionValid(order, version.getVersion()))
+            return new ResponseEntity<>(HttpStatus.CONFLICT);
+        orderRepository.save(order);
+        return new ResponseEntity<>(convertToDTO(order), HttpStatus.OK);
+    }
```
Comment (emiliskiskis) [inline geles_backend/src/main/java/lt/aigen/geles/controller/OrdersController.java:163]: I'd use braces for single line `if`s and also a bit more whitespace

File: geles_backend/src/main/java/lt/aigen/geles/models/User.java [hunk 21]
```diff
@@ -40,12 +40,16 @@
...
     private Cart cart;
     
-    @OneToMany(mappedBy = "user")
+    @OneToMany(mappedBy = "user", cascade = CascadeType.ALL, orphanRemoval = true)
     private List<CartTemplate> cartTemplates = new ArrayList<>();
 
+    @OneToMany(mappedBy = "user", cascade = CascadeType.ALL, orphanRemoval = true)
+    private List<Order> orders = new ArrayList<>();
+
+    @NotBlank
```
Comment (emiliskiskis) [inline geles_backend/src/main/java/lt/aigen/geles/models/User.java:52]: `NotBlank` fails on `Boolean`

File: geles_backend/src/main/java/lt/aigen/geles/models/FlowerInOrder.java [hunk 22]
```diff
@@ -24,7 +24,6 @@
     @JoinColumn(name = "order_id")
     private Order order;
 
-    @ManyToOne
```
Comment (emiliskiskis) [inline geles_backend/src/main/java/lt/aigen/geles/models/FlowerInOrder.java:27]: Mistake?

File: geles_backend/src/main/java/lt/aigen/geles/models/dto/OrderEditDTO.java [hunk 23]
```diff
@@ -0,0 +1,17 @@
+package lt.aigen.geles.models.dto;
+
+import lombok.Getter;
+import lombok.Setter;
+import lt.aigen.geles.models.Order;
+
+import java.util.Date;
```
Comment (emiliskiskis) [inline geles_backend/src/main/java/lt/aigen/geles/models/dto/OrderEditDTO.java:7]: Unused import

File: geles-frontend/src/ManageOrders.svelte [hunk 24]
```diff
@@ -0,0 +1,260 @@
+<script>
+  import { onMount } from "svelte";
+  import { user } from "./stores";
+  import axios from "axios";
+  import { Link } from "svelte-routing";
+  import { OrderStatus } from "./enums";
+  //import {Order, Flower} from "../types/flower";
```
Comment (emiliskiskis) [inline geles-frontend/src/ManageOrders.svelte:7]: Remove commented out code

File: geles-frontend/src/ManageOrders.svelte [hunk 25]
```diff
@@ -0,0 +1,260 @@
...
+
+  $: isAdmin = $user && $user.admin;
+
+  enum SortBy {
+    ORDER_ID,
+    ORDER_DATE,
+    ORDER_STATUS,
+    ORDER_USER
+  }
```
Comment (emiliskiskis) [inline geles-frontend/src/ManageOrders.svelte:16]: Seems unnecessary to specify `ORDER_`

**Review by emiliskiskis — CHANGES_REQUESTED** [review]
Shared problems between EditOrder.svelte and ManageOrders.svelte, all comments are on EditOrder.svelte. Avoid using `==` except for `== null` checks for unexpected undefined value.

File: geles-frontend/src/App.svelte (hunk 2 above)
Comment (CD10h) [reply geles-frontend/src/App.svelte:92]: done

File: geles-frontend/src/EditOrder.svelte (hunk 18 above)
Comment (CD10h) [reply geles-frontend/src/EditOrder.svelte:516]: what?

File: geles-frontend/types/flower.d.ts (hunk 19 above)
Comment (CD10h) [reply geles-frontend/types/flower.d.ts:34]: removing it breaks pages where it is used

Comment (emiliskiskis) [reply geles-frontend/types/flower.d.ts]: Then reimport the enum from enums.ts file

**Review by emiliskiskis — APPROVED** [review]
Amazing work!
